### ISSUE TITLE
Typing infrastructure improvements

### ISSUE THREAD

**Issue by PokhodenkoSA** [IntelPython/numba-dpex#143]
The basic goals are:
1. Add support for all SYCL types that can be passed as kernel arguments (skip buffers/accessors for now):
  - [ ] https://github.com/IntelPython/dpctl/issues/71 NumPy dtype and cython types to dpctl dtype
  - [ ] https://github.com/IntelPython/dpctl/issues/72 reduced/half precision types (i.e. BF16)
  - [x] https://github.com/IntelPython/dpctl/issues/85 Add support for complex numbers.
  - [ ] #1288
  - [ ] Array as type (w/o extraction)
2. Reduce the multiple places where we have checks to identify the kernel argument type:
  - [ ] _unpack_argument(): Numba to ctypes: for DPPYKernel
    https://github.com/IntelPython/numba-dppy/blob/2d100c5524e62b1fbd56ca9a96b173a75cf45bec/numba_dppy/compiler.py#L456
  - [ ] _populate_args(): ctypes to void** and DPCTLKernelArgType*: for submit
    https://github.com/IntelPython/dpctl/blob/609c8fd719ed61670127ad7c4e1dcbe783538ad6/dpctl/_sycl_queue.pyx#L116
  - [ ] set_kernel_arg(): void** and DPCTLKernelArgType* to set_arg(): for SYCL
    https://github.com/IntelPython/dpctl/blob/0763b213e1c1b5508f36964be5083a1ebea7ccaf/dpctl-capi/source/dpctl_sycl_queue_interface.cpp#L52
  - [ ] resolve_and_return_dpctl_type(): Numba type to DPCTLKernelArgType: for prange offload
    https://github.com/IntelPython/numba-dppy/blob/3e05b689f0eb7c67cb8d9b3fd0ec8bcfad316969/numba_dppy/dppy_host_fn_call_gen.py#L161

Tasks:
- [x] Evaluate SYCL/DPC++ types (Sec 4.16 of [SYCL 2020 Provisional Specification])
- [x] Evaluate the implementation of Ctypes types library
- [x] Evaluate the implementation for CUDA
  - https://github.com/IntelPython/numba/blob/5a1b372956ce881ebaf98c4ddd65681ece90ba0e/numba/cuda/compiler.py#L692
  - https://github.com/IntelPython/numba/blob/5a1b372956ce881ebaf98c4ddd65681ece90ba0e/numba/roc/compiler.py#L373
- [ ] Evaluate Numba - ctypes connection
  - https://github.com/IntelPython/numba/blob/5a1b372956ce881ebaf98c4ddd65681ece90ba0e/numba/core/typing/ctypes_utils.py#L62
- [ ] Provide a wrapper for SYCL/DPC++ types (Sec 4.16 of [SYCL 2020 Provisional Specification]) similar to what Ctypes does for C types

Design requirements:
1. Create a typing library inside dpctl similar to `сtypes._SimpleCData` or even deriving from it to have wrappers for all SYCL types including reduced precision.
2. Add new **types to Numba** as needed (e.g. for reduced precision types and short vector).
3. Have Numba-dppy specific type inference so that we can infer types as dpctl types.
4. Pass dpctl types to backend C API and do the type casting only once there.

Links:
1. [SYCL 2020 Provisional Specification]
2. https://github.com/IntelPython/dpctl/projects/1 - board

[SYCL 2020 Provisional Specification]: https://www.khronos.org/registry/SYCL/specs/sycl-2020-provisional.pdf

Problems:
1. How to support vector types? How to define vector types in Python? Vector - fixed size array.
2. DPC++ LLVM IR - For kernel getting vector. Is it possible to generate the same IR for vectors?

**Comment by PokhodenkoSA** [comment]
Data types we have:
1. Numba - numba.core.types
2. Ctypes - ctypes
3. DPCTLKernelArgType - enum
4. SYCL types which partially are ctypes

Numba -> Ctypes
DPPY. numba_dppy/compiler.py:DPPYKernel._unpack_argument - kernelargs (ctypes with values)
Numba. numba/core/typing/ctypes_utils.py - Numba <-> Ctypes conversion in both ways

Ctypes -> DPCTLKernelArgType
dpCtl.Cyhton. 
1. sycl_queue.submit -> _populate_args - Split to 2 arrays: values pointers and DPCTLKernelArgType types
2. sycl_queue.submit -> DPCTLQueue_SubmitRange

DPCTLKernelArgType -> SYCL types
dpCtl.backend. DPCTLQueue_SubmitRange -> set_kernel_arg - Dispatch for call handler cgh.set_arg(..., SYCL_TYPE)

**Comment by diptorupd** [comment]
Reduced precision types are not planned at this point of time. I have opened a new ticket #1288 to track the discussion around SYCL vector type support in our front end API.

The rest of the task list around `ctypes` is now moot as the kernel is going to be switched to the Numba `Dispatcher` API and will use native Numba boxing and unboxing and the use of ctypes will be removed.
